### PR TITLE
make the root scope responsible for periodic reporting. make better m…

### PR DESCRIPTION
**Pull request by jra3**
This adds special logic to NewScope that creates a registry of scopes/subscopes and attaches a timer loop that will dump the stats from the scopes to the StatsReporter.

I decided to update the example to illustrate how this api is to be used, and to detach it from the cactus library (real implementations for statsd should live in some other project)

### CODE REVIEW

File: scope.go [hunk 1]
```diff
@@ -23,12 +23,14 @@ package tally
 import (
 	"fmt"
 	"sync"
+	"time"
 
 	"github.com/facebookgo/clock"
 )
 
 type scope interface {
```
Comment (martin-mao) [inline scope.go]: I don't see where this interface is used?

File: scope.go [hunk 2]
```diff
@@ -103,6 +129,21 @@ func (s *standardScope) Report(r StatsReporter) {
 	// we do nothing for timers here because timers report directly to ths StatsReporter without buffering
 }
 
+// reportLoop is used by the root scope for periodic reporting
+func (s *standardScope) reportLoop(interval time.Duration) {
+	ticker := time.NewTicker(interval)
```
Comment (martin-mao) [inline scope.go]: Might be a good idea to close a Scope so we don't leak go-routines once we don't need the scope any more (ie in testing etc)

File: scope.go (hunk 1 above)
Comment (jra3) [inline scope.go]: "scope" is included in the "Scope" interface. It contains the non-exported methods

File: example/main.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	"fmt"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+type printStatsReporter struct{}
+
+// NewCactusStatsReporter returns a new StatsReporter
+func newPrintStatsReporter() tally.StatsReporter {
```
Comment (robskillington) [inline example/main.go]: Perhaps delete the comment or rename to reflect the name of the function "newPrintStatsReporter"

File: scope.go (hunk 2 above)
Comment (jra3) [inline scope.go]: Yeah, you have a point

Comment (robskillington) [inline scope.go]: Agreed, a scope.Close() method would be ideal.

File: scope.go [hunk 4]
```diff
@@ -103,6 +129,21 @@ func (s *standardScope) Report(r StatsReporter) {
...
+	for {
+		select {
+		case <-ticker.C:
+			s.registry.sm.Lock()
```
Comment (robskillington) [inline scope.go]: Do you need a write lock here?  Can you get away with just a RLock() and RUnlock()?

File: scope.go [hunk 5]
```diff
@@ -87,7 +84,36 @@ type standardScope struct {
...
+		tags = make(map[string]string)
+	}
+
+	scope := &standardScope{
+		prefix:   prefix,
+		tags:     tags,
+		reporter: reporter,
+
+		registry: &scopeRegistry{
+			subscopes: make([]Scope, 0, 5),
```
Comment (robskillington) [inline scope.go]: Five is kind of a strange number, perhaps a bigger number like 4096 or just don't allocate anything upfront at all for the default allocator?

File: scope.go (hunk 4 above)
Comment (jra3) [inline scope.go]: So we are either writing with registry.add, or we are reading from reportLoop. reportLoop is only called from one goroutine, so there's no point in using an RWMutex.

File: scope.go (hunk 5 above)
Comment (jra3) [inline scope.go]: 5 is the best number. I've got all the best numbers

File: scope.go [hunk 6]
```diff
@@ -21,12 +21,18 @@
...
+	"errors"
 	"fmt"
 	"sync"
+	"time"
 
 	"github.com/facebookgo/clock"
 )
 
+var (
+	errCloneNonRoot = errors.New("Called Close on non-root Scope")
```
Comment (robskillington) [inline scope.go]: We could create a new interface called RootScope which inherits Scope to avoid this perhaps?

Comment (jra3) [inline scope.go]: 👍

**Comment by robskillington** [comment]
LGTM

**Comment by martin-mao** [comment]
LGTM

**Pull request merged**